### PR TITLE
Translation module

### PR DESCRIPTION
**Pull request by Daiki4529**
Added a translation module, easy to edit and to add languages.

English and French are now supported.

To do :
- PGN, Captcha & generate report errors are missing (api.ts)
- Privacy Policy translation needed

### CODE REVIEW

File: src/public/pages/report/scripts/analysis.ts [hunk 1]
```diff
@@ -162,11 +162,11 @@ async function evaluate() {
         if (!positions.some((pos) => !pos.topLines)) {
             clearInterval(stockfishManager);
 
-            logAnalysisInfo("Evaluation complete.");
+            logAnalysisInfo(translateIntoUserLanguage(userLanguage,"Evaluation complete."));
             $("#evaluation-progress-bar").val(100);
             $(".g-recaptcha").css("display", "inline");
             $("#secondary-message").html(
-                "Please complete the CAPTCHA to continue.",
+                translateIntoUserLanguage(userLanguage,"Please complete the CAPTCHA to continue."),
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/analysis.ts:169]: Space after comma

File: src/public/pages/report/scripts/analysis.ts [hunk 2]
```diff
@@ -229,7 +229,7 @@ async function report() {
     $(".g-recaptcha").css("display", "none");
     $("#secondary-message").html("");
     $("#evaluation-progress-bar").attr("value", null);
-    logAnalysisInfo("Generating report...");
+    logAnalysisInfo(translateIntoUserLanguage(userLanguage,"Generating report..."));
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/analysis.ts:232]: Space after comma

File: src/public/pages/report/scripts/loadgame.ts [hunk 3]
```diff
@@ -49,7 +49,7 @@ async function fetchChessComGames(username: string) {
 
         let games: any[] = (await gamesResponse.json()).games;
 
-        $("#games-list").html(games.length == 0 ? "No games found." : "");
+        $("#games-list").html(games.length == 0 ? translateIntoUserLanguage(userLanguage,"No games found.") : "");
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/loadgame.ts:52]: Space after comma

File: src/public/pages/report/scripts/loadgame.ts [hunk 4]
```diff
@@ -68,7 +68,7 @@ async function fetchChessComGames(username: string) {
             $("#games-list").append(gameListing);
         }
     } catch {
-        $("#games-list").html("No games found.");
+        $("#games-list").html(translateIntoUserLanguage(userLanguage,"No games found."));
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/loadgame.ts:71]: Space after comma

File: src/public/pages/report/scripts/loadgame.ts [hunk 5]
```diff
@@ -100,7 +100,7 @@ async function fetchLichessGames(username: string) {
             .filter(game => game.length > 0)
             .map(game => JSON.parse(game));
 
-        $("#games-list").html(games.length == 0 ? "No games found." : "");
+        $("#games-list").html(games.length == 0 ? translateIntoUserLanguage(userLanguage,"No games found.") : "");
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/loadgame.ts:103]: Space after comma

File: src/public/pages/report/scripts/loadgame.ts [hunk 6]
```diff
@@ -119,7 +119,7 @@ async function fetchLichessGames(username: string) {
             $("#games-list").append(gameListing);
         }
     } catch {
-        $("#games-list").html("No games found.");
+        $("#games-list").html(translateIntoUserLanguage(userLanguage,"No games found."));
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/loadgame.ts:122]: Space after comma

File: src/public/pages/report/scripts/report.ts [hunk 7]
```diff
@@ -10,18 +10,21 @@ function updateClassificationMessage(lastPosition: Position, position: Position)
...
-            "inaccuracy": "an inaccuracy",
-            "mistake": "a mistake",
-            "blunder": "a blunder",
-            "book": "theory"
+            "great": translateIntoUserLanguage(userLanguage,"a great move"),
+            "good": translateIntoUserLanguage(userLanguage,"an okay move"),
+            "inaccuracy": translateIntoUserLanguage(userLanguage,"an inaccuracy"),
+            "mistake": translateIntoUserLanguage(userLanguage,"a mistake"),
+            "blunder": translateIntoUserLanguage(userLanguage,"a blunder"),
+            "book": translateIntoUserLanguage(userLanguage,"theory"),
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/report.ts:18]: Space after comma

File: src/public/pages/report/scripts/report.ts [hunk 8]
```diff
@@ -10,18 +10,21 @@ function updateClassificationMessage(lastPosition: Position, position: Position)
...
-            "good":"an okay move",
-            "inaccuracy": "an inaccuracy",
-            "mistake": "a mistake",
-            "blunder": "a blunder",
-            "book": "theory"
+            "great": translateIntoUserLanguage(userLanguage,"a great move"),
+            "good": translateIntoUserLanguage(userLanguage,"an okay move"),
+            "inaccuracy": translateIntoUserLanguage(userLanguage,"an inaccuracy"),
+            "mistake": translateIntoUserLanguage(userLanguage,"a mistake"),
+            "blunder": translateIntoUserLanguage(userLanguage,"a blunder"),
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/report.ts:17]: Space after comma

File: src/public/pages/report/scripts/report.ts [hunk 9]
```diff
@@ -10,18 +10,21 @@ function updateClassificationMessage(lastPosition: Position, position: Position)
...
-            "great": "a great move",
-            "good":"an okay move",
-            "inaccuracy": "an inaccuracy",
-            "mistake": "a mistake",
-            "blunder": "a blunder",
-            "book": "theory"
+            "great": translateIntoUserLanguage(userLanguage,"a great move"),
+            "good": translateIntoUserLanguage(userLanguage,"an okay move"),
+            "inaccuracy": translateIntoUserLanguage(userLanguage,"an inaccuracy"),
+            "mistake": translateIntoUserLanguage(userLanguage,"a mistake"),
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/report.ts:16]: Space after comma

File: src/public/pages/report/scripts/report.ts [hunk 10]
```diff
@@ -10,18 +10,21 @@ function updateClassificationMessage(lastPosition: Position, position: Position)
...
         let classificationMessages: { [key: string]: string } = {
-            "great": "a great move",
-            "good":"an okay move",
-            "inaccuracy": "an inaccuracy",
-            "mistake": "a mistake",
-            "blunder": "a blunder",
-            "book": "theory"
+            "great": translateIntoUserLanguage(userLanguage,"a great move"),
+            "good": translateIntoUserLanguage(userLanguage,"an okay move"),
+            "inaccuracy": translateIntoUserLanguage(userLanguage,"an inaccuracy"),
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/report.ts:15]: Space after comma

File: src/public/pages/report/scripts/report.ts [hunk 11]
```diff
@@ -10,18 +10,21 @@ function updateClassificationMessage(lastPosition: Position, position: Position)
...
     if (position.classification) {
         let classificationMessages: { [key: string]: string } = {
-            "great": "a great move",
-            "good":"an okay move",
-            "inaccuracy": "an inaccuracy",
-            "mistake": "a mistake",
-            "blunder": "a blunder",
-            "book": "theory"
+            "great": translateIntoUserLanguage(userLanguage,"a great move"),
+            "good": translateIntoUserLanguage(userLanguage,"an okay move"),
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/report.ts:14]: Space after comma

File: src/public/pages/report/scripts/report.ts [hunk 12]
```diff
@@ -10,18 +10,21 @@ function updateClassificationMessage(lastPosition: Position, position: Position)
 
     if (position.classification) {
         let classificationMessages: { [key: string]: string } = {
-            "great": "a great move",
-            "good":"an okay move",
-            "inaccuracy": "an inaccuracy",
-            "mistake": "a mistake",
-            "blunder": "a blunder",
-            "book": "theory"
+            "great": translateIntoUserLanguage(userLanguage,"a great move"),
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/report.ts:13]: Space after comma

File: src/public/pages/report/scripts/report.ts [hunk 13]
```diff
@@ -32,7 +35,7 @@ function updateClassificationMessage(lastPosition: Position, position: Position)
             let topAlternative = lastPosition.topLines?.[0].moveSAN;
             if (!topAlternative) return;
 
-            $("#top-alternative-message").html(`Best was ${topAlternative}`);
+            $("#top-alternative-message").html(`${translateIntoUserLanguage(userLanguage,"Best was")} ${topAlternative}`);
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/report.ts:38]: Space after comma

File: src/public/pages/report/scripts/translate.ts [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+                "placeholder",
+                placeholderTranslations[languageCode][translationKey]
+            );
+        }
+    });
+}
+
+const userLanguage = navigator.language.split("-")[0];
+
+translatePage(userLanguage);
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/translate.ts:77]: Probably you need to move function call to the bottom of file, after every function declaration. That's hard to see it here.

File: src/public/pages/report/scripts/translate.ts [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        ) {
+            $(element).attr(
+                "placeholder",
+                placeholderTranslations[languageCode][translationKey]
+            );
+        }
+    });
+}
+
+const userLanguage = navigator.language.split("-")[0];
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/translate.ts:75]: That const should be or at the start of file (you can easily find every const there when opening your file) or at the end (right before place where you use it).

File: src/public/pages/report/scripts/translate.ts [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+            "Failed to parse invalid PGN.": "Impossible de parser un PGN invalide.",
+            "PGN contains illegal moves.": "Le PGN contient des coups illégaux.",
+            "Failed to parse PGN.": "Impossible de parser le PGN.",
+            "Provide a game to analyse.": "Fournissez une partie à analyser.",
+            "Parsing PGN...": "Analyse du PGN...",
+        },
+    };
+    return translations[languageCode][key] ?? key;
+}
+document.addEventListener("click", () => translatePage(userLanguage));
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/translate.ts:143]: Please add some space after function declaration. Also you have no newline at the end of file.
More: https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

File: src/public/pages/report/scripts/report.ts [hunk 17]
```diff
@@ -10,18 +10,21 @@ function updateClassificationMessage(lastPosition: Position, position: Position)
...
-            "book": "theory"
+            "great": translateIntoUserLanguage(userLanguage,"a great move"),
+            "good": translateIntoUserLanguage(userLanguage,"an okay move"),
+            "inaccuracy": translateIntoUserLanguage(userLanguage,"an inaccuracy"),
+            "mistake": translateIntoUserLanguage(userLanguage,"a mistake"),
+            "blunder": translateIntoUserLanguage(userLanguage,"a blunder"),
+            "book": translateIntoUserLanguage(userLanguage,"theory"),
+            "best": translateIntoUserLanguage(userLanguage, "best"),
+            "brilliant": translateIntoUserLanguage(userLanguage, "brilliant"),
+            "forced": translateIntoUserLanguage(userLanguage, "forced")
```
Comment (Levsha1337) [inline src/public/pages/report/scripts/report.ts:21]: You already using dangling comma on other objects so you probably want it to be here aswell

**Review by Levsha1337 — COMMENTED** [review]
You need some codestyle fixes.
By the way, why didn't you just use the i18n module? It's easy to use, quite lightweight and it is bicycle-that-already-invented.
https://www.npmjs.com/package/i18n

**Comment by WintrCat** [comment]
preferably no dangling commas on any of the objects

File: src/public/pages/report/scripts/report.ts (hunk 17 above)
Comment (Levsha1337) [reply src/public/pages/report/scripts/report.ts:21]: Not necessary anymore

**Comment by Levsha1337** [comment]
Won't re-do review but there was a lot of dangling commas...

**Comment by MrMxffin** [comment]
Have you heard of i18n?

**Comment by WintrCat** [comment]
Stale

**Pull request closed without merging**